File: packages/snow_draw_core/lib/draw/render/tasks/render_tasks.dart
import 'package:meta/meta.dart';

import '../../config/draw_config.dart';
import '../../elements/core/element_data.dart';
import '../../elements/types/arrow/arrow_data.dart';
import '../../elements/types/arrow/arrow_points.dart';
import '../../elements/types/filter/filter_data.dart';
import '../../elements/types/free_draw/free_draw_data.dart';
import '../../elements/types/highlight/highlight_data.dart';
import '../../elements/types/line/line_data.dart';
import '../../elements/types/rectangle/rectangle_data.dart';
import '../../elements/types/serial_number/serial_number_data.dart';
import '../../elements/types/text/text_data.dart';
import '../../models/element_state.dart';
import '../../types/draw_color.dart';
import '../../types/draw_point.dart';
import '../../types/draw_rect.dart';
import '../../types/snap_guides.dart';

/// Base type for backend-executable render tasks.
sealed class RenderTask {
  const RenderTask();
}

/// Base type for element render tasks produced by core.
@immutable
sealed class ElementRenderTask<T extends ElementData> extends RenderTask {
  const ElementRenderTask({
    required this.element,
    required this.data,
    this.localeTag,
  });

  /// Effective element state in world-space.
  final ElementState element;

  /// Typed element payload.
  final T data;

  /// Optional locale hint used by text-capable backends.
  final String? localeTag;
}

/// Rectangle rendering task.
@immutable
final class RectangleRenderTask extends ElementRenderTask<RectangleData> {
  const RectangleRenderTask({
    required super.element,
    required super.data,
    super.localeTag,
  });
}

/// Line rendering task.
@immutable
final class LineRenderTask extends ElementRenderTask<LineData> {
  const LineRenderTask({
    required super.element,
    required super.data,
    super.localeTag,
  });
}

/// Arrow rendering task.
@immutable
final class ArrowRenderTask extends ElementRenderTask<ArrowData> {
  const ArrowRenderTask({
    required super.element,
    required super.data,
    super.localeTag,
  });
}

/// Free-draw rendering task.
@immutable
final class FreeDrawRenderTask extends ElementRenderTask<FreeDrawData> {
  const FreeDrawRenderTask({
    required super.element,
    required super.data,
    super.localeTag,
  });
}

/// Text rendering task.
@immutable
final class TextRenderTask extends ElementRenderTask<TextData> {
  const TextRenderTask({
    required super.element,
    required super.data,
    super.localeTag,
  });
}

/// Serial-number rendering task.
@immutable
final class SerialNumberRenderTask extends ElementRenderTask<SerialNumberData> {
  const SerialNumberRenderTask({
    required super.element,
    required super.data,
    super.localeTag,
  });
}

/// Highlight rendering task.
@immutable
final class HighlightRenderTask extends ElementRenderTask<HighlightData> {
  const HighlightRenderTask({
    required super.element,
    required super.data,
    super.localeTag,
  });
}

/// Filter rendering task.
@immutable
final class FilterRenderTask extends ElementRenderTask<FilterData> {
  const FilterRenderTask({
    required super.element,
    required super.data,
    super.localeTag,
  });
}

/// Background paint task.
@immutable
final class BackgroundRenderTask extends RenderTask {
  const BackgroundRenderTask({required this.color});

  final DrawColor color;
}

/// Grid paint task.
@immutable
final class GridRenderTask extends RenderTask {
  const GridRenderTask({
    required this.enabled,
    required this.size,
    required this.lineWidth,
    required this.lineColor,
    required this.lineOpacity,
    required this.majorLineEvery,
    required this.majorLineOpacity,
    required this.minScreenSpacing,
    required this.minRenderSpacing,
  });

  final bool enabled;
  final double size;
  final double lineWidth;
  final DrawColor lineColor;
  final double lineOpacity;
  final int majorLineEvery;
  final double majorLineOpacity;
  final double minScreenSpacing;
  final double minRenderSpacing;
}

/// Selection-outline task.
@immutable
final class SelectionOutlineRenderTask extends RenderTask {
  const SelectionOutlineRenderTask({
    required this.bounds,
    required this.config,
    this.rotation,
    this.rotationCenter,
    this.dashed = true,
  });

  final DrawRect bounds;
  final SelectionConfig config;
  final double? rotation;
  final DrawPoint? rotationCenter;
  final bool dashed;
}

/// Selection controls task (outline + handles).
@immutable
final class SelectionControlsRenderTask extends RenderTask {
  const SelectionControlsRenderTask({
    required this.bounds,
    required this.config,
    this.rotation,
    this.rotationCenter,
    this.dashed = true,
    this.cornerHandleOffset = 0.0,
    this.showRotationHandle = true,
  });

  final DrawRect bounds;
  final SelectionConfig config;
  final double? rotation;
  final DrawPoint? rotationCenter;
  final bool dashed;
  final double cornerHandleOffset;
  final bool showRotationHandle;
}

/// Arrow-point overlay task.
@immutable
final class ArrowPointOverlayRenderTask extends RenderTask {
  const ArrowPointOverlayRenderTask({
    required this.handles,
    required this.selectionConfig,
    this.activeHandle,
    this.hoveredHandle,
    this.deleteIndicatorVisible = false,
  });

  final List<ArrowPointHandle> handles;
  final SelectionConfig selectionConfig;
  final ArrowPointHandle? activeHandle;
  final ArrowPointHandle? hoveredHandle;
  final bool deleteIndicatorVisible;
}

/// Hover-outline task.
@immutable
final class HoverOutlineRenderTask extends RenderTask {
  const HoverOutlineRenderTask({
    required this.element,
    required this.config,
    this.useTextUnderlineStyle = false,
  });

  final ElementState element;
  final SelectionConfig config;
  final bool useTextUnderlineStyle;
}

/// Snap-guides overlay task.
@immutable
final class SnapGuidesRenderTask extends RenderTask {
  const SnapGuidesRenderTask({required this.guides, required this.snapConfig});

  final List<SnapGuide> guides;
  final SnapConfig snapConfig;
}

/// Box-select overlay task.
@immutable
final class BoxSelectionRenderTask extends RenderTask {
  const BoxSelectionRenderTask({
    required this.bounds,
    required this.config,
    this.previewElements = const <ElementState>[],
  });

  final DrawRect bounds;
  final BoxSelectionConfig config;
  final List<ElementState> previewElements;
}

/// Highlight-mask overlay task.
@immutable
final class HighlightMaskRenderTask extends RenderTask {
  const HighlightMaskRenderTask({
    required this.config,
    required this.highlights,
  });

  final HighlightMaskConfig config;
  final List<ElementState> highlights;
}

/// Watermark overlay task.
@immutable
final class WatermarkRenderTask extends RenderTask {
  const WatermarkRenderTask({required this.config});

  final WatermarkConfig config;
}

File: packages/snow_draw_rust_engine/src/draw/render/tasks/render_tasks.rs
#![allow(dead_code)]

use crate::draw::config::canvas_config::BoxSelectionConfig;
use crate::draw::config::draw_config::SelectionConfig;
use crate::draw::config::highlight_config::HighlightMaskConfig;
use crate::draw::config::snap_config::SnapConfig;
use crate::draw::config::watermark_config::WatermarkConfig;
use crate::draw::elements::types::arrow::arrow_data::ArrowData;
use crate::draw::elements::types::connector::connector_points::ConnectorPointHandle;
use crate::draw::elements::types::filter::filter_data::FilterData;
use crate::draw::elements::types::free_draw::free_draw_data::FreeDrawData;
use crate::draw::elements::types::highlight::highlight_data::HighlightData;
use crate::draw::elements::types::line::line_data::LineData;
use crate::draw::elements::types::rectangle::rectangle_data::RectangleData;
use crate::draw::elements::types::serial_number::serial_number_data::SerialNumberData;
use crate::draw::elements::types::text::text_data::TextData;
use crate::draw::models::draw_state_view::ElementState;
use crate::draw::types::draw_color::DrawColor;
use crate::draw::types::draw_point::DrawPoint;
use crate::draw::types::draw_rect::DrawRect;
use crate::draw::types::snap_guides::SnapGuide;

/// Base type for backend-executable render tasks.
#[derive(Clone, Debug, PartialEq)]
pub enum RenderTask {
    Frame(FrameRenderTask),
    Element(ElementTask),
}

/// Base type for frame-level tasks consumed by the scene painter.
#[derive(Clone, Debug, PartialEq)]
pub enum FrameRenderTask {
    Background(BackgroundRenderTask),
    Grid(GridRenderTask),
    SelectionOutline(SelectionOutlineRenderTask),
    SelectionControls(SelectionControlsRenderTask),
    ConnectorPointOverlay(ConnectorPointOverlayRenderTask),
    ArrowPointOverlay(ArrowPointOverlayRenderTask),
    ArrowBindingHighlight(ArrowBindingHighlightRenderTask),
    HoverOutline(HoverOutlineRenderTask),
    SnapGuides(SnapGuidesRenderTask),
    BoxSelection(BoxSelectionRenderTask),
    HighlightMask(HighlightMaskRenderTask),
    Watermark(WatermarkRenderTask),
}

/// Base type for element render tasks produced by the engine.
#[derive(Clone, Debug, PartialEq)]
pub enum ElementTask {
    Rectangle(RectangleRenderTask),
    Line(LineRenderTask),
    Arrow(ArrowRenderTask),
    FreeDraw(FreeDrawRenderTask),
    Text(TextRenderTask),
    SerialNumber(SerialNumberRenderTask),
    Highlight(HighlightRenderTask),
    Filter(FilterRenderTask),
}

/// Shared payload for typed element rendering tasks.
#[derive(Clone, Debug, PartialEq)]
pub struct ElementRenderTask<T> {
    /// Effective element state in world-space.
    pub element: ElementState,
    /// Typed element payload.
    pub data: T,
    /// Optional locale hint used by text-capable backends.
    pub locale_tag: Option<String>,
}

impl<T> ElementRenderTask<T> {
    pub fn new(element: ElementState, data: T, locale_tag: Option<String>) -> Self {
        Self {
            element,
            data,
            locale_tag,
        }
    }
}

pub type RectangleRenderTask = ElementRenderTask<RectangleData>;
pub type LineRenderTask = ElementRenderTask<LineData>;
pub type ArrowRenderTask = ElementRenderTask<ArrowData>;
pub type FreeDrawRenderTask = ElementRenderTask<FreeDrawData>;
pub type TextRenderTask = ElementRenderTask<TextData>;
pub type SerialNumberRenderTask = ElementRenderTask<SerialNumberData>;
pub type HighlightRenderTask = ElementRenderTask<HighlightData>;
pub type FilterRenderTask = ElementRenderTask<FilterData>;

/// Background paint task.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub struct BackgroundRenderTask {
    pub color: DrawColor,
}

impl BackgroundRenderTask {
    pub const fn new(color: DrawColor) -> Self {
        Self { color }
    }
}

/// Grid paint task.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct GridRenderTask {
    pub enabled: bool,
    pub size: f64,
    pub line_width: f64,
    pub line_color: DrawColor,
    pub line_opacity: f64,
    pub major_line_every: i64,
    pub major_line_opacity: f64,
    pub min_screen_spacing: f64,
    pub min_render_spacing: f64,
}

/// Selection-outline task.
#[derive(Clone, Debug, PartialEq)]
pub struct SelectionOutlineRenderTask {
    pub bounds: DrawRect,
    pub config: SelectionConfig,
    pub rotation: Option<f64>,
    pub rotation_center: Option<DrawPoint>,
    pub dashed: bool,
}

/// Selection controls task (outline + handles).
#[derive(Clone, Debug, PartialEq)]
pub struct SelectionControlsRenderTask {
    pub bounds: DrawRect,
    pub config: SelectionConfig,
    pub rotation: Option<f64>,
    pub rotation_center: Option<DrawPoint>,
    pub dashed: bool,
    pub corner_handle_offset: f64,
    pub show_rotation_handle: bool,
}

/// Connector-point overlay task.
#[derive(Clone, Debug, PartialEq)]
pub struct ConnectorPointOverlayRenderTask {
    pub handles: Vec<ConnectorPointHandle>,
    pub selection_config: SelectionConfig,
    pub active_handle: Option<ConnectorPointHandle>,
    pub hovered_handle: Option<ConnectorPointHandle>,
    pub delete_indicator_visible: bool,
}

pub type ArrowPointOverlayRenderTask = ConnectorPointOverlayRenderTask;

/// Arrow-binding highlight task.
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct ArrowBindingHighlightRenderTask {
    pub element_ids: Vec<String>,
    pub stroke_color: DrawColor,
}

/// Hover-outline task.
#[derive(Clone, Debug, PartialEq)]
pub struct HoverOutlineRenderTask {
    pub element: ElementState,
    pub config: SelectionConfig,
    pub use_text_underline_style: bool,
}

/// Snap-guides overlay task.
#[derive(Clone, Debug, PartialEq)]
pub struct SnapGuidesRenderTask {
    pub guides: Vec<SnapGuide>,
    pub snap_config: SnapConfig,
}

/// Box-select overlay task.
#[derive(Clone, Debug, PartialEq)]
pub struct BoxSelectionRenderTask {
    pub bounds: DrawRect,
    pub config: BoxSelectionConfig,
    pub selection_config: SelectionConfig,
    pub preview_elements: Vec<ElementState>,
}

/// Highlight-mask overlay task.
#[derive(Clone, Debug, PartialEq)]
pub struct HighlightMaskRenderTask {
    pub config: HighlightMaskConfig,
    pub highlights: Vec<ElementState>,
}

/// Watermark overlay task.
#[derive(Clone, Debug, PartialEq)]
pub struct WatermarkRenderTask {
    pub config: WatermarkConfig,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::draw::config::draw_config::SelectionConfig;

    #[test]
    fn connector_overlay_variant_preserves_payload() {
        let task = ConnectorPointOverlayRenderTask {
            handles: Vec::new(),
            selection_config: SelectionConfig::default(),
            active_handle: None,
            hovered_handle: None,
            delete_indicator_visible: true,
        };

        let frame_task = FrameRenderTask::ConnectorPointOverlay(task.clone());

        assert_eq!(frame_task, FrameRenderTask::ConnectorPointOverlay(task));
    }

    #[test]
    fn legacy_arrow_overlay_alias_still_builds_variant() {
        let task = ArrowPointOverlayRenderTask {
            handles: Vec::new(),
            selection_config: SelectionConfig::default(),
            active_handle: None,
            hovered_handle: None,
            delete_indicator_visible: false,
        };

        let frame_task = FrameRenderTask::ArrowPointOverlay(task.clone());

        assert_eq!(frame_task, FrameRenderTask::ArrowPointOverlay(task));
    }
}
